### PR TITLE
Use gc values in connectors

### PR DESCRIPTION
**Pull request by do4gr**
Refactoring to ease the implementation of new connectors.
Sets up a typed interface for the dataresolvers and cleans up data conversions within Prisma Core.

### CODE REVIEW

File: server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseQueryBuilder.scala [hunk 1]
```diff
@@ -332,46 +214,48 @@ object DatabaseQueryBuilder {
...
+      .as[PrismaNodeWithParent](getResultForModelAndRelationSide(relatedModel, fromField.relationSide.get.toString))
+      .map { items =>
+        val itemGroupsByModelId = items.groupBy(_.parentId)
+        fromModelIds
+          .map(id =>
+            itemGroupsByModelId.find(_._1 == id) match {
+              case Some((_, itemsForId)) => args.get.resultTransform(itemsForId).copy(parentModelId = Some(id))
+              case None                  => ResolverResult(Vector.empty[PrismaNodeWithParent], hasPreviousPage = false, hasNextPage = false, parentModelId = Some(id))
+          })
+      }
```
Comment (sorenbs) [inline server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseQueryBuilder.scala]: Could we get rid of the map/map/match nesting?

File: server/connectors/api-connector/src/main/scala/com/prisma/api/connector/DataResolverTypes.scala [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 
 import com.prisma.api.connector.Types.DataItemFilterCollection
+import com.prisma.gc_values.IdGCValue
 import com.prisma.shared.models.{Field, Model, Relation}
 
 import scala.collection.immutable.Seq
 
 object Types {
-  type DataItemFilterCollection = Seq[_ >: Seq[Any] <: Any]
+  type DataItemFilterCollection = Seq[_ >: Seq[Any] <: Any] //todo
```
Comment (sorenbs) [inline server/connectors/api-connector/src/main/scala/com/prisma/api/connector/DataResolverTypes.scala]: remember todo :-)

File: server/connectors/deploy-connector-mysql/src/main/scala/com/prisma/deploy/connector/mysql/database/DatabaseMutationBuilder.scala [hunk 3]
```diff
@@ -55,8 +55,8 @@ object DatabaseMutationBuilder {
   }
 
   def updateScalarListType(projectId: String, modelName: String, fieldName: String, typeIdentifier: TypeIdentifier) = {
-    val sqlType       = sqlTypeForScalarTypeIdentifier(false, typeIdentifier)
-    val charsetString = charsetTypeForScalarTypeIdentifier(false, typeIdentifier)
+    val sqlType = sqlTypeForScalarTypeIdentifier(false, typeIdentifier)
+//    val charsetString = charsetTypeForScalarTypeIdentifier(false, typeIdentifier)
```
Comment (sorenbs) [inline server/connectors/deploy-connector-mysql/src/main/scala/com/prisma/deploy/connector/mysql/database/DatabaseMutationBuilder.scala]: do we need this?

File: server/libs/gc-values/src/main/scala/com/prisma/gc_values/GcValues.scala [hunk 4]
```diff
@@ -15,38 +16,80 @@ import scala.collection.immutable.SortedMap
...
 case class RootGCValue(map: SortedMap[String, GCValue]) extends GCValue {
   def idField = map.get("id") match {
-    case Some(id) => id.asInstanceOf[GraphQLIdGCValue]
+    case Some(id) => id.asInstanceOf[IdGCValue]
     case None     => sys.error("There was no field with name 'id'.")
   }
 
   def filterValues(p: GCValue => Boolean) = copy(map = map.filter(t => p(t._2)))
+
+//  def toJson: JsValue = GCValueExtractor.fromGCValueToJson(this)
```
Comment (sorenbs) [inline server/libs/gc-values/src/main/scala/com/prisma/gc_values/GcValues.scala]: do we need this?

File: server/servers/api/src/main/scala/com/prisma/api/mutactions/DatabaseMutactions.scala [hunk 5]
```diff
@@ -1,94 +1,74 @@
...
+    Vector(VerifyWhere(project, path.root)) ++ generateCascadingDeleteMutactions(path) ++
+      Vector(DeleteRelationCheck(project, path), DeleteDataItem(project, path, previousValues))
   }
 
-  def getMutactionsForUpdate(path: Path, args: CoolArgs, id: Id, previousValues: DataItem): Seq[DatabaseMutaction] = report {
-    val updateMutaction = getUpdateMutactions(path, args, id, previousValues)
-    val nested          = getMutactionsForNestedMutation(args, path.updatedRoot(args), triggeredFromCreate = false)
-
-    updateMutaction ++ nested
+  //todo this does not support cascading delete behavior at the moment
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/mutactions/DatabaseMutactions.scala]: remember todo

File: server/servers/api/src/main/scala/com/prisma/api/mutactions/DatabaseMutactions.scala [hunk 6]
```diff
@@ -1,94 +1,74 @@
...
+    val (nonListArgs, listArgs) = args.getCreateArgs(path)
+    val createMutaction         = CreateDataItem(project, path, nonListArgs, listArgs)
+    val nestedMutactions        = getMutactionsForNestedMutation(args, path, triggeredFromCreate = true)
 
-    updateNonLists +: updateLists
+    createMutaction +: nestedMutactions
   }
 
-  def getSetScalarList(path: Path, field: Field, values: Vector[Any]): SetScalarList = SetScalarList(project, path, field, values)
+  // todo this still needs to implement execution of nested mutactions
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/mutactions/DatabaseMutactions.scala]: Does nested mutations currently not work in upsert or is this a new deficit?

File: server/servers/api/src/main/scala/com/prisma/api/resolver/ToOneDeferredResolver.scala [hunk 7]
```diff
@@ -17,47 +18,39 @@ class ToOneDeferredResolver(dataResolver: DataResolver) {
...
 
-    def matchesRelation(dataItem: DataItem, relationSide: String) =
-      dataItem.userData
-        .get(relationSide)
-        .flatten
-        .contains(deferred.parentNodeId)
+    def matchesRelation(prismaNodeWithParent: PrismaNodeWithParent, relationSide: String) =
+      prismaNodeWithParent.parentId == deferred.parentNodeId
 
     // see https://github.com/graphcool/internal-docs/blob/master/relations.md#findings
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/resolver/ToOneDeferredResolver.scala]: remove comment

File: server/servers/api/src/main/scala/com/prisma/api/schema/FilterArguments.scala [hunk 8]
```diff
@@ -71,22 +71,22 @@ object FilterArguments {
...
-    FilterArgument("_length", "All values matching the given length."),
-    FilterArgument("_length_not", "All values not matching the given length."),
-    FilterArgument("_length_lt", "All values with a length less than the given length."),
-    FilterArgument("_length_lte", "All values with a length less than or equal the given length."),
-    FilterArgument("_length_gt", "All values with a length greater than the given length."),
-    FilterArgument("_length_gte", "All values with a length less than or equal the given length."),
-    FilterArgument("_length_in", "All values that have one of the lengths specified."),
-    FilterArgument("_length_not_in", "All values that do not have any of the lengths specified.")
-  )
+//  private val listFilters = List(
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/schema/FilterArguments.scala]: Let's remember to delete this

File: server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala [hunk 9]
```diff
@@ -261,131 +254,50 @@ class ObjectTypeBuilder(
...
-        ToOneDeferred(field, item.id, arguments)
-      }
-
-    } else {
-      if (field.isList) {
-        ScalarListDeferred(model, field, item.id)
-      } else {
-        ObjectTypeBuilder.convertScalarFieldValueFromDatabase(field, item)
-      }
+    (field.isScalar, field.isList) match {
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala]: nice 👍

File: server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala [hunk 10]
```diff
@@ -261,131 +254,50 @@ class ObjectTypeBuilder(
...
-                    }
-                }
-              case TypeIdentifier.Enum => mapTo[String](value)
-              case TypeIdentifier.Json => mapTo[JsValue](value)
-            }
-        }
+  def convertScalarFieldValueFromDatabase(field: models.Field, item: PrismaNode): Any = {
+    import com.prisma.util.json.PlaySprayConversions._
+
+    //todo use GCValueextractor once we got rid of Spray
```
Comment (sorenbs) [inline server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala]: Do you have an issue to remove Spray where we track things we can clean up then?

File: server/servers/api/src/test/scala/com/prisma/util/gcvalueconverters/GCDBStringEndToEndSpec.scala [hunk 11]
```diff
@@ -1,122 +0,0 @@
-package com.prisma.util.gcvalueconverters
```
Comment (sorenbs) [inline server/servers/api/src/test/scala/com/prisma/util/gcvalueconverters/GCDBStringEndToEndSpec.scala]: Let's make sure we reintroduce these checks

File: server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala [hunk 12]
```diff
@@ -106,13 +106,15 @@ object DatabaseMutationBuilder {
              create: DBIOAction[Any, NoStream, Effect],
              update: DBIOAction[Any, NoStream, Effect]) = {
 
-    val q = (sql"select exists" ++ DatabaseMutationBuilder.pathQueryForLastChild(projectId, updatePath)).as[Boolean]
+    val query = sql"select exists ( SELECT `id` FROM `#$projectId`.`#${updatePath.lastModel.name}` WHERE `id` = " ++ pathQueryForLastChild(projectId,
```
Comment (sorenbs) [inline server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala]: pathQueryForLastChild Can either return a sub query or the id. This means that sometimes this code will generate an extra sub query. We need to clean this up later.

File: server/connectors/deploy-connector-mysql/src/main/scala/com/prisma/deploy/connector/mysql/database/DatabaseMutationBuilder.scala (hunk 3 above)
Comment (do4gr) [reply server/connectors/deploy-connector-mysql/src/main/scala/com/prisma/deploy/connector/mysql/database/DatabaseMutationBuilder.scala]: not there, moved it

File: server/libs/gc-values/src/main/scala/com/prisma/gc_values/GcValues.scala (hunk 4 above)
Comment (do4gr) [reply server/libs/gc-values/src/main/scala/com/prisma/gc_values/GcValues.scala]: removed it

File: server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala (hunk 10 above)
Comment (do4gr) [reply server/servers/api/src/main/scala/com/prisma/api/schema/ObjectTypeBuilder.scala]: Opened One

File: server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala (hunk 12 above)
Comment (do4gr) [reply server/connectors/api-connector-mysql/src/main/scala/com/prisma/api/connector/mysql/database/DatabaseMutationBuilder.scala]: Yeah, the next pass over the path logic should look into when we can simplify paths by looking for ids and when they are fully necessary to check relations. I tried to separate the checks from the paths, but once everything is cleaned up we can make sure we generate less complex queries wherever possible

**Pull request merged**